### PR TITLE
Improve `unknown` narrowing by negated type predicates

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/60789

### CODE REVIEW

**Comment by gabritto** [comment]
@typescript-bot test it

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -29404,8 +29404,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 if (checkDerived) {
                     return filterType(type, t => !isTypeDerivedFrom(t, candidate));
                 }
+                const originalType = type;
+                type = type.flags & TypeFlags.Unknown ? unknownUnionType : type;
                 const trueType = getNarrowedType(type, candidate, /*assumeTrue*/ true, /*checkDerived*/ false);
-                return filterType(type, t => !isTypeSubsetOf(t, trueType));
+                const filtered = filterType(type, t => !isTypeSubsetOf(t, trueType));
+                return originalType.flags & TypeFlags.Unknown && filtered === type ? originalType : filtered;
```
Comment (gabritto) [inline src/compiler/checker.ts:29411]: Maybe we can use `recombineUnknownType` here?

Comment (Andarist) [reply src/compiler/checker.ts:29411]: Good point, done

**Comment by gabritto** [comment]
@typescript-bot pack this

**Review by gabritto — APPROVED** [review]

**Comment by gabritto** [comment]
I'm now slightly curious about what it would take to fix this example: https://www.typescriptlang.org/play/?ts=5.8.0-dev.20241218#code/GYVwdgxgLglg9mABAIwIYBMA8AVRBTADyjzHQGdFwBrMOAdzAD4AKAgLkWwEoOA3OGOkQBvAFCJEMYIlaIAhAF5KpPMBhg86LiPETEEBGSiIAnh2EBfREoIBuXRd14ANmTw69+w8bOIwIZ2dEAB9ldFV1TWtEO0QAejjEAFEAJxS4FI4AAwIsyQpaYzBUNPoogAs8FLwHUUdRUEhYBEQAczg4LFxCYlIKSxC-AKDQ8HC1DXQWdk4eRH5BD0lpWUUwiMntMU8DMCNTcysbewlHCRc3JYld-d9-QMGxjajj+MSAdQyqMmzc-L8Suk6BUqjVTnUgA
I used it as an example to justify why you should use union types explicitly in https://github.com/microsoft/TypeScript/pull/56941, but at some point I might have to update that example to not use `unknown`, as that is a special type we sometimes know to break into a union.

**Pull request merged**